File: 06-Arrays-Hashing/04-Kadanes-Algorithm/cpp/max_subarray.cpp
/*
 * Problem: Maximum Subarray (Kadane's Algorithm)
 * Source:  LeetCode 53
 * Time: O(n), Space: O(1)
 * Tags: kadanes, dp, array
 */
#include <bits/stdc++.h>
using namespace std;

int maxSubArray(vector<int>& nums) {
    int maxSum = nums[0], curSum = nums[0];
    for (int i = 1; i < nums.size(); i++) {
        curSum = max(nums[i], curSum + nums[i]);
        maxSum = max(maxSum, curSum);
    }
    return maxSum;
}

int main() {
    vector<int> nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << "Max Subarray Sum: " << maxSubArray(nums) << endl; // 6
    return 0;
}

File: 06-Arrays-Hashing/04-Kadanes-Algorithm/java/MaxSubarray.java
public class MaxSubarray {
    public static int maxSubArray(int[] nums) {
        int maxSum = nums[0], curSum = nums[0];
        for (int i = 1; i < nums.length; i++) {
            curSum = Math.max(nums[i], curSum + nums[i]);
            maxSum = Math.max(maxSum, curSum);
        }
        return maxSum;
    }
    public static void main(String[] args) {
        System.out.println("Max Subarray Sum: " + maxSubArray(new int[]{-2,1,-3,4,-1,2,1,-5,4})); // 6
    }
}
